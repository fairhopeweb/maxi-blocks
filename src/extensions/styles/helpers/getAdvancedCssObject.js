const trimUnmatchedBrace = code => {
	const braceIndex = code.indexOf('{');
	if (braceIndex !== -1) {
		const lastSemicolonBeforeBrace = code.lastIndexOf(';', braceIndex);
		return code.substring(0, lastSemicolonBeforeBrace + 1).trim();
	}
	return code;
};

const getAdvancedCssObject = obj => {
	const response = {};
	const code = obj['advanced-css'];

	if (!code) return response;

	const selectorRegex = /([a-zA-Z0-9\-_\s.,#:*[\]="']*?)\s*{([^}]*)}/g;

	let remainingCode = code;
	let match = selectorRegex.exec(code);

	while (match) {
		const rawSelectors = match[1]?.trim();
		const properties = trimUnmatchedBrace(match[2]?.trim());

		if (properties && !properties.includes('{')) {
			// Split selectors by comma and create separate response entries for each selector
			rawSelectors.split(',').forEach(rawSelector => {
				const selector = ` ${rawSelector.trim()}`;
				response[selector] = {
					advancedCss: {
						general: {
							css: properties,
						},
					},
				};
			});

			remainingCode = remainingCode.replace(match[0], '').trim(); // Remove the parsed segment from the remaining code
		} else {
			// if unmatched brace is found, stop the loop to prevent endless loop scenario
			break;
		}

		match = selectorRegex.exec(code);
	}

	// Add the remaining part as general CSS
	if (remainingCode) {
		remainingCode = trimUnmatchedBrace(remainingCode);

		if (response['']) {
			response[''].advancedCss.general.css += `\n${remainingCode}`;
		} else {
			response[''] = {
				advancedCss: {
					general: {
						css: remainingCode,
					},
				},
			};
		}
	}

	return response;
};

export default getAdvancedCssObject;
