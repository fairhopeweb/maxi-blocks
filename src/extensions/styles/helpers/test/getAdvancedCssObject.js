import getAdvancedCssObject from '../getAdvancedCssObject';

describe('getAdvancedCssObject', () => {
	it('should handle basic CSS correctly', () => {
		const input = {
			'advanced-css': `
				background: red;
				.maxi-block-button {
					color: yellow;
				}
			`,
		};
		const result = getAdvancedCssObject(input);
		expect(result).toMatchSnapshot();
	});

	it('should handle CSS with no selectors', () => {
		const input = {
			'advanced-css': 'background: blue;',
		};
		const result = getAdvancedCssObject(input);
		expect(result).toMatchSnapshot();
	});

	it('should handle CSS with multiple selectors', () => {
		const input = {
			'advanced-css': `
				background: green;
				.maxi-block-button {
					color: yellow;
				}
				.maxi-block-button,
				.maxi-block-button:hover {
					color: red;
				}
			`,
		};

		const result = getAdvancedCssObject(input);
		expect(result).toMatchSnapshot();
	});

	it('should handle CSS with pseudo selectors', () => {
		const input = {
			'advanced-css': `
				background: green;
				.maxi-block-button {
					color: yellow;
				}
				.maxi-block-button:hover {
					color: red;
				}
				.maxi-block-button::before {
					content: 'before';
				}
			`,
		};
		const result = getAdvancedCssObject(input);
		expect(result).toMatchSnapshot();
	});

	// Edge/invalid cases
	it('should handle CSS with missing closing brace', () => {
		const input = {
			'advanced-css': `
				background: yellow;
				p {
					text-align: center;
			`,
		};
		const result = getAdvancedCssObject(input);
		expect(result).toMatchSnapshot();
	});

	it('should handle CSS with missing closing brace in nested braces', () => {
		const input = {
			'advanced-css': `
				background: yellow;
				p {
					text-align: center;

					div {
						color: red;
				}
			`,
		};
		const result = getAdvancedCssObject(input);
		expect(result).toMatchSnapshot();
	});

	it('should handle CSS with multiple unmatched braces', () => {
		const input = {
			'advanced-css': `
				background: teal;
				div {
					padding: 10px;
				p {
					margin: 5px;
			`,
		};
		const result = getAdvancedCssObject(input);
		expect(result).toMatchSnapshot();
	});

	it('should handle empty CSS correctly', () => {
		const input = {
			'advanced-css': '',
		};
		const result = getAdvancedCssObject(input);
		expect(result).toMatchSnapshot();
	});
});
