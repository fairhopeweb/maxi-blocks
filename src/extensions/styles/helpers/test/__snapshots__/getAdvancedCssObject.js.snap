// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAdvancedCssObject should handle CSS with missing closing brace 1`] = `
{
  "": {
    "advancedCss": {
      "general": {
        "css": "background: yellow;",
      },
    },
  },
}
`;

exports[`getAdvancedCssObject should handle CSS with missing closing brace in nested braces 1`] = `
{
  "": {
    "advancedCss": {
      "general": {
        "css": "background: yellow;",
      },
    },
  },
  " p": {
    "advancedCss": {
      "general": {
        "css": "text-align: center;",
      },
    },
  },
}
`;

exports[`getAdvancedCssObject should handle CSS with multiple selectors 1`] = `
{
  "": {
    "advancedCss": {
      "general": {
        "css": "background: green;",
      },
    },
  },
  " .maxi-block-button": {
    "advancedCss": {
      "general": {
        "css": "color: red;",
      },
    },
  },
  " .maxi-block-button:hover": {
    "advancedCss": {
      "general": {
        "css": "color: red;",
      },
    },
  },
}
`;

exports[`getAdvancedCssObject should handle CSS with multiple unmatched braces 1`] = `
{
  "": {
    "advancedCss": {
      "general": {
        "css": "background: teal;",
      },
    },
  },
}
`;

exports[`getAdvancedCssObject should handle CSS with no selectors 1`] = `
{
  "": {
    "advancedCss": {
      "general": {
        "css": "background: blue;",
      },
    },
  },
}
`;

exports[`getAdvancedCssObject should handle CSS with pseudo selectors 1`] = `
{
  "": {
    "advancedCss": {
      "general": {
        "css": "background: green;",
      },
    },
  },
  " .maxi-block-button": {
    "advancedCss": {
      "general": {
        "css": "color: yellow;",
      },
    },
  },
  " .maxi-block-button::before": {
    "advancedCss": {
      "general": {
        "css": "content: 'before';",
      },
    },
  },
  " .maxi-block-button:hover": {
    "advancedCss": {
      "general": {
        "css": "color: red;",
      },
    },
  },
}
`;

exports[`getAdvancedCssObject should handle basic CSS correctly 1`] = `
{
  "": {
    "advancedCss": {
      "general": {
        "css": "background: red;",
      },
    },
  },
  " .maxi-block-button": {
    "advancedCss": {
      "general": {
        "css": "color: yellow;",
      },
    },
  },
}
`;

exports[`getAdvancedCssObject should handle empty CSS correctly 1`] = `{}`;
